File: UtilWeDie-Neo-1.21.1/ModsSourceCode/ZombieSurvivalKit/ZSK-1.21.1-neoforge/src/main/java/com/scarasol/zombiekit/client/model/MortarModel.java
package com.scarasol.zombiekit.client.model;

import com.scarasol.zombiekit.ZombieKitMod;
import com.scarasol.zombiekit.entity.mechanics.MortarEntity;
import net.minecraft.resources.ResourceLocation;
import net.minecraft.util.Mth;
import software.bernie.geckolib.constant.DataTickets;
import software.bernie.geckolib.cache.object.GeoBone;
import software.bernie.geckolib.animation.AnimationState;
import software.bernie.geckolib.model.GeoModel;
import software.bernie.geckolib.model.data.EntityModelData;

public class MortarModel extends GeoModel<MortarEntity> {
    @Override
    public ResourceLocation getAnimationResource(MortarEntity entity) {
        return ResourceLocation.fromNamespaceAndPath(ZombieKitMod.MODID, "animations/mortar.animation.json");
    }

    @Override
    public ResourceLocation getModelResource(MortarEntity entity) {
        return ResourceLocation.fromNamespaceAndPath(ZombieKitMod.MODID, "geo/" + entity.getModel() + ".geo.json");
    }

    @Override
    public ResourceLocation getTextureResource(MortarEntity entity) {
        return ResourceLocation.fromNamespaceAndPath(ZombieKitMod.MODID, "textures/entities/mortar.png");
    }

    @Override
    public void setCustomAnimations(MortarEntity animatable, long instanceId, AnimationState<MortarEntity> animationState) {
        GeoBone head = getAnimationProcessor().getBone("head");
        if (head != null) {
            EntityModelData entityData = animationState.getData(DataTickets.ENTITY_MODEL_DATA);
            head.setRotX((entityData.headPitch() - 45) * Mth.DEG_TO_RAD);
            head.setRotY(entityData.netHeadYaw() * Mth.DEG_TO_RAD);
        }

    }
}

File: SocialWill/sw_src/src/main/kotlin/com/atsuishio/superbwarfare/client/model/entity/MortarModel.kt
package com.atsuishio.superbwarfare.client.model.entity

import com.atsuishio.superbwarfare.entity.vehicle.MortarEntity
import net.minecraft.util.Mth
import software.bernie.geckolib.constant.DataTickets

class MortarModel : VehicleModel<MortarEntity>() {
    override fun collectTransform(boneName: String): TransformContext<MortarEntity>? {
        if (boneName == "paoguan") {
            return TransformContext { bone, _, state ->
                val jiaojia = animationProcessor.getBone("jiaojia")
                val entityData = state.getData(DataTickets.ENTITY_MODEL_DATA)
                if (entityData != null) {
                    bone.rotX = (entityData.headPitch()) * Mth.DEG_TO_RAD
                    jiaojia.rotX =
                        -2 * ((entityData.headPitch() - (10 - entityData.headPitch() * 0.1f)) * Mth.DEG_TO_RAD)
                }
            }
        }

        if (boneName == "monitor") {
            return TransformContext { bone, vehicle, _ ->
                bone.isHidden = !vehicle.getEntityData().get(MortarEntity.INTELLIGENT)
            }
        }

        return super.collectTransform(boneName)
    }
}
